p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,275 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ tpt.CapableConn = &connection{}
+
+const (
+	maxAcceptQueueLen int = 10
+)
+
+type connection struct {
+	pc        *webrtc.PeerConnection
+	transport *WebRTCTransport
+	scope     network.ConnManagementScope
+
+	localPeer      peer.ID
+	privKey        ic.PrivKey
+	localMultiaddr ma.Multiaddr
+
+	remotePeer      peer.ID
+	remoteKey       ic.PubKey
+	remoteMultiaddr ma.Multiaddr
+
+	m       sync.Mutex
```

Comment:
This is still the case. Consider using a generic solution (developed yourself,  or used from the community) where you are forced to use this safely.

E.g.:

```go
type concurrentSafeStreams struct {
  mux sync.Mutex
  streams map[uint16]*webRTCStream
}

func (c *concurrentSafeStreams) addStream(id uint16, stream *webRTCStream) {
   c.mux.lock()
   // NOTE: even though this protects the shallow map here,
   // it does not provide any guarantees for the (*webRTCStream) assigned
   // to be use for concurrent use, as it might as well be used concurrently.
   // if that is not safe, consider creating or copying it first!
   c.streams[id] = stream
   c.mux.unlock()
}

// adapt the callback structure and method to your own needs obviously,
// if needed at all
func (c* concurrentSafeStreams) forEachStream(fn func(id uint16, stream *webRTCStream)) {
   c.mux.lock()
   defer c.mux.unlock()
   for id, stream := range c.streams {
      fn(id, stream)
    }
}
```

The above is just an example, but the goal is just to make sure that you always protect your value actually by the lock. Your approach works as well, but is a lot harder to reason about and requires more discipline. It is therefore recommended to either use a safe generic wrapping structure safe for concurrent use or make a specialised one yourself.